Add CMake support

p12tic opened the pull request:
This adds CMakeLists.txt from vcpkg. The project compiles locally, I haven't yet tested it with AliceVision. I will add AliceVision support in a subsequent PR.

p12tic commented:
@simogasp We should probably have `av_develop` in this repository too.

simogasp commented:
> @simogasp We should probably have `av_develop` in this repository too.

done, it's the default branch now and I changed the target branch of this PR as well

simogasp commented on CMakeLists.txt:
```diff
@@ -0,0 +1,182 @@
+# SOFTWARE.
+
+cmake_minimum_required (VERSION 3.10)
+project (vlfeat)
```
```suggestion
project (vlfeat LANGUAGES C VERSION 0.9.21)
```
The latest tag 0.9.21 points at the latest commit on the master branch, so I guess we can start from there,

p12tic replied:
Fixed now, thanks.

simogasp commented on CMakeLists.txt:
```diff
@@ -0,0 +1,182 @@
+    "Debug" "Release" "MinSizeRel" "RelWithDebInfo")
+endif()
+
+if(ENABLE_OPENMP)
```
I'd add an option like this at the beginning
```cmake
option(VLFEAT_ENABLE_OPENMP "Enable OpenMP optimization" ON)
```
the `VLFEAT_` is just to avoid ambiguities

p12tic replied:
Fixed now, thanks.

simogasp commented on CMakeLists.txt:
```diff
@@ -0,0 +1,182 @@
+
+if(ENABLE_OPENMP)
+  find_package(OpenMP REQUIRED)
+  set (CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${OpenMP_C_FLAGS}")
```
I think we can remove this if we link to the provided target `OpenMP::OpenMP_C`

p12tic replied:
Yes, this is oversight from my refactoring. Fixed now.

simogasp commented on CMakeLists.txt:
```diff
@@ -0,0 +1,182 @@
+  set (CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${OpenMP_C_FLAGS}")
+endif()
+
+if(USE_SSE)
```
same here, maybe add an option at the beginning
```cmake
option(VLFEAT_USE_SSE "Enable SSE optimizations" ON)
option(VLFEAT_USE_AVX "Enable AVX optimization" ON)
```

p12tic replied:
Fixed now, thanks.

simogasp commented on CMakeLists.txt:
```diff
@@ -0,0 +1,182 @@
+    )
+endif()
+
+target_compile_options(vl PRIVATE ${OpenMP_C_FLAGS})
```
I think we can replace this with 
```suggestion
if(VLFEAT_ENABLE_OPENMP)
    target_link_libraries(vl PRIVATE OpenMP::OpenMP_C)
endif()
```
that should import everything is needed, i think.
I haven't checked, but unless openmp is used in some hpp it should be ok to set it as `PRIVATE`. Can you confirm?

p12tic replied:
Thanks, this makes sense. We can make it private, indeed.

p12tic replied:
Fixed now.

simogasp commented on CMakeLists.txt:
```diff
@@ -0,0 +1,182 @@
+  FILE vlfeatConfig.cmake
+  NAMESPACE vlfeat::
+  DESTINATION "${INSTALL_CMAKE_DIR}"
+)
```
It's better to add the version file as well
```cmake
include(CMakePackageConfigHelpers)

set(version_config "${CMAKE_CURRENT_BINARY_DIR}/${PROJECT_NAME}ConfigVersion.cmake")

write_basic_package_version_file("${version_config}"
                                 COMPATIBILITY SameMajorVersion)

install(FILES ${version_config}
        DESTINATION "${INSTALL_CMAKE_DIR}")
```

p12tic replied:
Fixed now, thanks.

simogasp reviewed (changes requested):
Thanks for the PR!
I left some comments

p12tic committed:
Add cmake configuration from vcpkg

p12tic committed:
Use standard install paths for CMake files

p12tic committed:
Don't change input CMAKE_C_FLAGS

p12tic committed:
Clean up compile options setup in CMake

p12tic committed:
Add Project version to CMake

p12tic committed:
Add explicit CMake options for controllable behavior

p12tic committed:
Add CMake version config

p12tic committed:
Disable OpenMP support by default in CMake

At the moment OpenMP support does not compile cleanly due to obvious
issues like "omp parallel" having "default(none)" clause and then using
variables from enclosing scope.

p12tic commented:
@simogasp I have fixed all issues that you've identified. Note, that I disable OpenMP support because the current code is broken and needs investigation on whether the simple compile issues I've seen are the only ones and there isn't more significant bitrot.

p12tic commented:
@simogasp Just a friendly ping :-)

simogasp commented:
thanks for the reminder, I've forgotten
